File: web/src/lib/graph/Graph.ts
/**
 * Graph Database Implementation
 *
 * Inspired by NeuVector's in-memory graph database
 * Provides a lightweight directed multigraph with O(1) operations
 */

import {
  GraphNode,
  GraphLink,
  NewLinkCallback,
  DelNodeCallback,
  DelLinkCallback,
  UpdateLinkAttrCallback,
  ConnectedNodeCallback,
  GraphStats,
  GraphExportOptions,
  D3GraphData,
  CytoscapeElement,
} from './types';

/**
 * Main Graph class - manages nodes and edges with callbacks
 */
export class Graph<TAttr = any> {
  private nodes: Map<string, GraphNode<TAttr>> = new Map();
  private cbNewLink?: NewLinkCallback;
  private cbDelNode?: DelNodeCallback;
  private cbDelLink?: DelLinkCallback;
  private cbUpdateLinkAttr?: UpdateLinkAttrCallback;

  constructor() {
    this.nodes = new Map();
  }

  /**
   * Register callback for new link creation
   */
  registerNewLinkCallback(cb: NewLinkCallback): void {
    this.cbNewLink = cb;
  }

  /**
   * Register callback for node deletion
   */
  registerDelNodeCallback(cb: DelNodeCallback): void {
    this.cbDelNode = cb;
  }

  /**
   * Register callback for link deletion
   */
  registerDelLinkCallback(cb: DelLinkCallback): void {
    this.cbDelLink = cb;
  }

  /**
   * Register callback for link attribute update
   */
  registerUpdateLinkAttrCallback(cb: UpdateLinkAttrCallback): void {
    this.cbUpdateLinkAttr = cb;
  }

  /**
   * Add a directed edge from src to dst with given link type and attribute
   * O(1) operation
   * Creates nodes automatically if they don't exist
   */
  addLink(src: string, linkType: string, dst: string, attr: TAttr): void {
    // Ensure source node exists
    let srcNode = this.nodes.get(src);
    if (!srcNode) {
      srcNode = new GraphNode<TAttr>();
      this.nodes.set(src, srcNode);
    }

    // Ensure destination node exists
    let dstNode = this.nodes.get(dst);
    if (!dstNode) {
      dstNode = new GraphNode<TAttr>();
      this.nodes.set(dst, dstNode);
    }

    // Add outgoing link from src
    let outLink = srcNode.outs.get(linkType);
    if (!outLink) {
      outLink = new GraphLink<TAttr>();
      srcNode.outs.set(linkType, outLink);
    }

    const isNewLink = !outLink.ends.has(dst);
    outLink.ends.set(dst, attr);

    // Add incoming link to dst
    let inLink = dstNode.ins.get(linkType);
    if (!inLink) {
      inLink = new GraphLink<TAttr>();
      dstNode.ins.set(linkType, inLink);
    }
    inLink.ends.set(src, attr);

    // Trigger callbacks
    if (isNewLink && this.cbNewLink) {
      this.cbNewLink(src, linkType, dst);
    } else if (!isNewLink && this.cbUpdateLinkAttr) {
      this.cbUpdateLinkAttr(src, linkType, dst);
    }
  }

  /**
   * Delete a specific directed edge
   * O(1) operation
   */
  deleteLink(src: string, linkType: string, dst: string): void {
    const srcNode = this.nodes.get(src);
    const dstNode = this.nodes.get(dst);

    if (!srcNode || !dstNode) {
      return;
    }

    // Remove from source node's outgoing links
    const outLink = srcNode.outs.get(linkType);
    if (outLink) {
      outLink.ends.delete(dst);
      if (outLink.ends.size === 0) {
        srcNode.outs.delete(linkType);
      }
    }

    // Remove from destination node's incoming links
    const inLink = dstNode.ins.get(linkType);
    if (inLink) {
      inLink.ends.delete(src);
      if (inLink.ends.size === 0) {
        dstNode.ins.delete(linkType);
      }
    }

    // Trigger callback
    if (this.cbDelLink) {
      this.cbDelLink(src, linkType, dst);
    }
  }

  /**
   * Delete a node and all its associated edges
   * Cascades to remove all incoming and outgoing links
   */
  deleteNode(node: string): void {
    const nodeData = this.nodes.get(node);
    if (!nodeData) {
      return;
    }

    // Delete all outgoing links
    for (const [linkType, link] of nodeData.outs.entries()) {
      for (const dst of link.ends.keys()) {
        this.deleteLink(node, linkType, dst);
      }
    }

    // Delete all incoming links
    for (const [linkType, link] of nodeData.ins.entries()) {
      for (const src of link.ends.keys()) {
        this.deleteLink(src, linkType, node);
      }
    }

    // Remove the node itself
    this.nodes.delete(node);

    // Trigger callback
    if (this.cbDelNode) {
      this.cbDelNode(node);
    }
  }

  /**
   * Get the attribute of a specific edge
   * Returns undefined if edge doesn't exist
   */
  attr(src: string, linkType: string, dst: string): TAttr | undefined {
    const srcNode = this.nodes.get(src);
    if (!srcNode) {
      return undefined;
    }

    const outLink = srcNode.outs.get(linkType);
    if (!outLink) {
      return undefined;
    }

    return outLink.ends.get(dst);
  }

  /**
   * Get all incoming neighbors of a node for a specific link type
   * If linkType is not specified, returns all incoming neighbors across all link types
   * O(1) for specific link type, O(k) where k is number of link types otherwise
   */
  ins(node: string, linkType?: string): string[] {
    const nodeData = this.nodes.get(node);
    if (!nodeData) {
      return [];
    }

    if (linkType) {
      const link = nodeData.ins.get(linkType);
      return link ? Array.from(link.ends.keys()) : [];
    }

    // Get all incoming neighbors across all link types
    const neighbors = new Set<string>();
    for (const link of nodeData.ins.values()) {
      for (const neighbor of link.ends.keys()) {
        neighbors.add(neighbor);
      }
    }
    return Array.from(neighbors);
  }

  /**
   * Get all outgoing neighbors of a node for a specific link type
   * If linkType is not specified, returns all outgoing neighbors across all link types
   * O(1) for specific link type, O(k) where k is number of link types otherwise
   */
  outs(node: string, linkType?: string): string[] {
    const nodeData = this.nodes.get(node);
    if (!nodeData) {
      return [];
    }

    if (linkType) {
      const link = nodeData.outs.get(linkType);
      return link ? Array.from(link.ends.keys()) : [];
    }

    // Get all outgoing neighbors across all link types
    const neighbors = new Set<string>();
    for (const link of nodeData.outs.values()) {
      for (const neighbor of link.ends.keys()) {
        neighbors.add(neighbor);
      }
    }
    return Array.from(neighbors);
  }

  /**
   * Get all neighbors (both incoming and outgoing) of a node
   * If linkType is not specified, returns all neighbors across all link types
   */
  both(node: string, linkType?: string): string[] {
    const inNeighbors = this.ins(node, linkType);
    const outNeighbors = this.outs(node, linkType);

    const neighbors = new Set<string>([...inNeighbors, ...outNeighbors]);
    return Array.from(neighbors);
  }

  /**
   * Get all nodes in the graph
   */
  all(): string[] {
    return Array.from(this.nodes.keys());
  }

  /**
   * Check if a node exists in the graph
   */
  hasNode(node: string): boolean {
    return this.nodes.has(node);
  }

  /**
   * Get the number of nodes in the graph
   */
  getNodeCount(): number {
    return this.nodes.size;
  }

  /**
   * Get the total number of edges in the graph
   */
  getEdgeCount(): number {
    let count = 0;
    for (const node of this.nodes.values()) {
      for (const link of node.outs.values()) {
        count += link.ends.size;
      }
    }
    return count;
  }

  /**
   * Get all unique link types in the graph
   */
  getLinkTypes(): string[] {
    const linkTypes = new Set<string>();
    for (const node of this.nodes.values()) {
      for (const linkType of node.outs.keys()) {
        linkTypes.add(linkType);
      }
    }
    return Array.from(linkTypes);
  }

  /**
   * Get graph statistics
   */
  getStats(): GraphStats {
    const nodeCount = this.getNodeCount();
    const edgeCount = this.getEdgeCount();
    const linkTypes = this.getLinkTypes();
    const averageDegree = nodeCount > 0 ? (edgeCount * 2) / nodeCount : 0;

    return {
      nodeCount,
      edgeCount,
      linkTypes,
      averageDegree,
    };
  }

  /**
   * Clear all nodes and edges from the graph
   */
  clear(): void {
    this.nodes.clear();
  }

  /**
   * Export graph data in various formats
   */
  export(options: GraphExportOptions): any {
    const { format, includeAttributes = false, linkTypes } = options;

    switch (format) {
      case 'json':
        return this.exportJSON(includeAttributes, linkTypes);
      case 'd3':
        return this.exportD3(includeAttributes, linkTypes);
      case 'cytoscape':
        return this.exportCytoscape(includeAttributes, linkTypes);
      default:
        throw new Error(`Unsupported export format: ${format}`);
    }
  }

  /**
   * Export as raw JSON structure
   */
  private exportJSON(includeAttributes: boolean, linkTypes?: string[]): any {
    const result: any = {
      nodes: this.all(),
      edges: [],
    };

    for (const [src, srcNode] of this.nodes.entries()) {
      for (const [linkType, link] of srcNode.outs.entries()) {
        if (linkTypes && !linkTypes.includes(linkType)) {
          continue;
        }

        for (const [dst, attr] of link.ends.entries()) {
          const edge: any = {
            source: src,
            target: dst,
            type: linkType,
          };

          if (includeAttributes && attr !== undefined) {
            edge.attributes = attr;
          }

          result.edges.push(edge);
        }
      }
    }

    return result;
  }

  /**
   * Export in D3.js force graph format
   */
  private exportD3(includeAttributes: boolean, linkTypes?: string[]): D3GraphData {
    const nodes = this.all().map(id => ({ id }));
    const links: D3GraphData['links'] = [];

    for (const [src, srcNode] of this.nodes.entries()) {
      for (const [linkType, link] of srcNode.outs.entries()) {
        if (linkTypes && !linkTypes.includes(linkType)) {
          continue;
        }

        for (const [dst, attr] of link.ends.entries()) {
          const linkData: any = {
            source: src,
            target: dst,
            type: linkType,
          };

          if (includeAttributes && attr !== undefined) {
            Object.assign(linkData, attr);
          }

          links.push(linkData);
        }
      }
    }

    return { nodes, links };
  }

  /**
   * Export in Cytoscape.js format
   */
  private exportCytoscape(includeAttributes: boolean, linkTypes?: string[]): CytoscapeElement[] {
    const elements: CytoscapeElement[] = [];

    // Add nodes
    for (const node of this.all()) {
      elements.push({
        data: { id: node },
      });
    }

    // Add edges
    for (const [src, srcNode] of this.nodes.entries()) {
      for (const [linkType, link] of srcNode.outs.entries()) {
        if (linkTypes && !linkTypes.includes(linkType)) {
          continue;
        }

        for (const [dst, attr] of link.ends.entries()) {
          const edgeData: any = {
            id: `${src}-${linkType}-${dst}`,
            source: src,
            target: dst,
            type: linkType,
          };

          if (includeAttributes && attr !== undefined) {
            Object.assign(edgeData, attr);
          }

          elements.push({ data: edgeData });
        }
      }
    }

    return elements;
  }
}

File: src/server/pkg/graph/graph.go
// input: node and edge operations (add/delete/update)
// output: graph query results, adjacency lookups, connectivity analysis
// pos: graph - in-memory directed multi-graph data structure
//
// Package graph provides an in-memory graph database implementation.
// Adapted from NeuVector's controller/graph/graph.go (Apache 2.0 License).

package graph

import (
	"encoding/json"
	"reflect"
	"sync"
)

// Callback types for graph events.
type (
	NewLinkCallback        func(src, link, dst string)
	DelNodeCallback        func(node string)
	DelLinkCallback        func(src, link, dst string)
	UpdateLinkAttrCallback func(src, link, dst string)
	ConnectedNodeCallback  func(node string) bool
	PurgeOutLinkCallback   func(src, link, dst string, attr interface{}, param interface{}) bool
)

// graphLink represents edges leading to other nodes, indexed by node name.
type graphLink struct {
	ends map[string]interface{} // destination node name -> edge attribute
}

// graphNode represents a node with incoming and outgoing links indexed by link type.
type graphNode struct {
	ins  map[string]*graphLink // link type -> incoming edges
	outs map[string]*graphLink // link type -> outgoing edges
}

// Graph is an in-memory directed multi-graph with typed edges.
// It supports:
// - Multiple edge types between the same pair of nodes (via link name)
// - Arbitrary attributes on edges
// - Callback hooks for graph mutations
type Graph struct {
	mu               sync.RWMutex
	nodes            map[string]*graphNode
	cbNewLink        NewLinkCallback
	cbDelNode        DelNodeCallback
	cbDelLink        DelLinkCallback
	cbUpdateLinkAttr UpdateLinkAttrCallback
}

// NewGraph creates and returns a new empty Graph.
func NewGraph() *Graph {
	return &Graph{
		nodes: make(map[string]*graphNode),
	}
}

// RegisterNewLinkHook sets the callback for new link creation.
func (g *Graph) RegisterNewLinkHook(cb NewLinkCallback) {
	g.cbNewLink = cb
}

// RegisterDelNodeHook sets the callback for node deletion.
func (g *Graph) RegisterDelNodeHook(cb DelNodeCallback) {
	g.cbDelNode = cb
}

// RegisterDelLinkHook sets the callback for link deletion.
func (g *Graph) RegisterDelLinkHook(cb DelLinkCallback) {
	g.cbDelLink = cb
}

// RegisterUpdateLinkAttrHook sets the callback for link attribute updates.
func (g *Graph) RegisterUpdateLinkAttrHook(cb UpdateLinkAttrCallback) {
	g.cbUpdateLinkAttr = cb
}

// Reset clears all nodes and edges from the graph.
func (g *Graph) Reset() {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.nodes = make(map[string]*graphNode)
}

// AddLink adds or updates an edge from src to dst with the given link type and attribute.
// If the edge already exists and the attribute differs, it will be updated.
func (g *Graph) AddLink(src, link, dst string, attr interface{}) {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.addLinkUnsafe(src, link, dst, attr)
}

// addLinkUnsafe is the internal implementation without locking.
func (g *Graph) addLinkUnsafe(src, link, dst string, attr interface{}) {
	var gn *graphNode
	var gl *graphLink
	var ok, newlink, updattr bool

	// Handle source node's outgoing edge
	if gn, ok = g.nodes[src]; !ok {
		gl = &graphLink{ends: make(map[string]interface{})}
		gl.ends[dst] = attr

		gn = &graphNode{
			ins:  make(map[string]*graphLink),
			outs: make(map[string]*graphLink),
		}
		gn.outs[link] = gl

		g.nodes[src] = gn
		newlink = true
	} else if gl, ok = gn.outs[link]; !ok {
		gl = &graphLink{ends: make(map[string]interface{})}
		gl.ends[dst] = attr

		gn.outs[link] = gl
		newlink = true
	} else if _, ok = gl.ends[dst]; !ok {
		gl.ends[dst] = attr
		newlink = true
	} else {
		if !reflect.DeepEqual(gl.ends[dst], attr) {
			gl.ends[dst] = attr
			updattr = true
		}
	}

	// Handle destination node's incoming edge
	if gn, ok = g.nodes[dst]; !ok {
		gl = &graphLink{ends: make(map[string]interface{})}
		gl.ends[src] = attr

		gn = &graphNode{
			ins:  make(map[string]*graphLink),
			outs: make(map[string]*graphLink),
		}
		gn.ins[link] = gl

		g.nodes[dst] = gn
		newlink = true
	} else if gl, ok = gn.ins[link]; !ok {
		gl = &graphLink{ends: make(map[string]interface{})}
		gl.ends[src] = attr

		gn.ins[link] = gl
		newlink = true
	} else if _, ok = gl.ends[src]; !ok {
		gl.ends[src] = attr
		newlink = true
	} else {
		if !reflect.DeepEqual(gl.ends[src], attr) {
			gl.ends[src] = attr
			updattr = true
		}
	}

	if newlink && g.cbNewLink != nil {
		g.cbNewLink(src, link, dst)
	}
	if updattr && g.cbUpdateLinkAttr != nil {
		g.cbUpdateLinkAttr(src, link, dst)
	}
}

// Attr returns the attribute of the edge from src to dst with the given link type.
// Returns nil if the edge doesn't exist.
func (g *Graph) Attr(src, link, dst string) interface{} {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if s, ok := g.nodes[src]; ok {
		if gl, ok := s.outs[link]; ok {
			if attr, ok := gl.ends[dst]; ok {
				return attr
			}
		}
	}
	return nil
}

// DeleteLink removes the edge from src to dst with the given link type.
func (g *Graph) DeleteLink(src, link, dst string) {
	g.mu.Lock()
	defer g.mu.Unlock()
	g.deleteLinkUnsafe(src, link, dst)
}

// deleteLinkUnsafe is the internal implementation without locking.
func (g *Graph) deleteLinkUnsafe(src, link, dst string) {
	var s, d *graphNode
	var ok bool

	if s, ok = g.nodes[src]; !ok {
		return
	}
	if d, ok = g.nodes[dst]; !ok {
		return
	}

	if gl, ok := s.outs[link]; ok {
		if _, ok = gl.ends[dst]; ok {
			delete(gl.ends, dst)
			if len(gl.ends) == 0 {
				delete(s.outs, link)
				if g.cbDelLink != nil {
					g.cbDelLink(src, link, dst)
				}
			}
		}
	}

	if gl, ok := d.ins[link]; ok {
		if _, ok = gl.ends[src]; ok {
			delete(gl.ends, src)
			if len(gl.ends) == 0 {
				delete(d.ins, link)
			}
		}
	}
}

// DeleteNode removes a node and all its edges from the graph.
func (g *Graph) DeleteNode(node string) string {
	g.mu.Lock()
	defer g.mu.Unlock()

	var gn *graphNode
	var ok bool

	if gn, ok = g.nodes[node]; !ok {
		return ""
	}

	// Delete all incoming edges
	for link, gl := range gn.ins {
		for n := range gl.ends {
			g.deleteLinkUnsafe(n, link, node)
		}
	}

	// Delete all outgoing edges
	for link, gl := range gn.outs {
		for n := range gl.ends {
			g.deleteLinkUnsafe(node, link, n)
		}
	}

	delete(g.nodes, node)

	if g.cbDelNode != nil {
		g.cbDelNode(node)
	}

	return node
}

// Node returns the node name if it exists, empty string otherwise.
func (g *Graph) Node(v string) string {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[v]; ok {
		return v
	}
	return ""
}

// All returns a set of all node names in the graph.
func (g *Graph) All() Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := NewSet()
	for v := range g.nodes {
		ret.Add(v)
	}
	return ret
}

// GetNodeCount returns the total number of nodes.
func (g *Graph) GetNodeCount() int {
	g.mu.RLock()
	defer g.mu.RUnlock()
	return len(g.nodes)
}

// NoIn returns nodes with no incoming edges.
func (g *Graph) NoIn() Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := NewSet()
	for v, n := range g.nodes {
		if len(n.ins) == 0 {
			ret.Add(v)
		}
	}
	return ret
}

// NoInByLink returns nodes with no incoming edges of the specified link type.
func (g *Graph) NoInByLink(link string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := NewSet()
	for v, n := range g.nodes {
		if gl, ok := n.ins[link]; !ok || len(gl.ends) == 0 {
			ret.Add(v)
		}
	}
	return ret
}

// NoOut returns nodes with no outgoing edges.
func (g *Graph) NoOut() Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := NewSet()
	for v, n := range g.nodes {
		if len(n.outs) == 0 {
			ret.Add(v)
		}
	}
	return ret
}

// NoOutByLink returns nodes with no outgoing edges of the specified link type.
func (g *Graph) NoOutByLink(link string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := NewSet()
	for v, n := range g.nodes {
		if gl, ok := n.outs[link]; !ok || len(gl.ends) == 0 {
			ret.Add(v)
		}
	}
	return ret
}

// Ins returns all nodes with incoming edges to this node.
func (g *Graph) Ins(node string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[node]; !ok {
		return nil
	}

	ret := NewSet()
	n := g.nodes[node]
	for _, l := range n.ins {
		for v := range l.ends {
			ret.Add(v)
		}
	}
	return ret
}

// InsByLink returns all nodes with incoming edges of the specified link type.
func (g *Graph) InsByLink(node, link string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[node]; !ok {
		return nil
	}

	ret := NewSet()
	n := g.nodes[node]
	if gl, ok := n.ins[link]; ok {
		for v := range gl.ends {
			ret.Add(v)
		}
	}
	return ret
}

// Outs returns all nodes with outgoing edges from this node.
func (g *Graph) Outs(node string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[node]; !ok {
		return nil
	}

	ret := NewSet()
	n := g.nodes[node]
	for _, l := range n.outs {
		for v := range l.ends {
			ret.Add(v)
		}
	}
	return ret
}

// OutsByLink returns all nodes with outgoing edges of the specified link type.
func (g *Graph) OutsByLink(node, link string) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[node]; !ok {
		return nil
	}

	ret := NewSet()
	n := g.nodes[node]
	if gl, ok := n.outs[link]; ok {
		for v := range gl.ends {
			ret.Add(v)
		}
	}
	return ret
}

// Both returns all nodes connected to this node (incoming or outgoing).
func (g *Graph) Both(node string) Set {
	ins := g.Ins(node)
	outs := g.Outs(node)
	if ins == nil {
		return nil
	}
	return ins.Union(outs)
}

// BothByLink returns all nodes connected by the specified link type.
func (g *Graph) BothByLink(node, link string) Set {
	ins := g.InsByLink(node, link)
	outs := g.OutsByLink(node, link)
	if ins == nil {
		return nil
	}
	return ins.Union(outs)
}

// Connected returns all nodes reachable from the given node.
func (g *Graph) Connected(node string, cb ConnectedNodeCallback) Set {
	g.mu.RLock()
	defer g.mu.RUnlock()

	if _, ok := g.nodes[node]; !ok {
		return nil
	}

	ret := NewSet()
	ret.Add(node)
	q := []string{node}

	for len(q) > 0 {
		node, q = q[0], q[1:]

		both := g.bothUnsafe(node)
		for n := range both.Iter() {
			nodeStr := n.(string)
			if cb != nil && cb(nodeStr) {
				if !ret.Contains(n) {
					ret.Add(n)
					q = append(q, nodeStr)
				}
			}
		}
	}
	return ret
}

// bothUnsafe is internal implementation without locking.
func (g *Graph) bothUnsafe(node string) Set {
	ret := NewSet()
	if n, ok := g.nodes[node]; ok {
		for _, l := range n.ins {
			for v := range l.ends {
				ret.Add(v)
			}
		}
		for _, l := range n.outs {
			for v := range l.ends {
				ret.Add(v)
			}
		}
	}
	return ret
}

// BetweenDirLinks returns all link types and their attributes between src and dst.
func (g *Graph) BetweenDirLinks(src, dst string) map[string]interface{} {
	g.mu.RLock()
	defer g.mu.RUnlock()

	ret := make(map[string]interface{})
	if n, ok := g.nodes[src]; ok {
		for ln, l := range n.outs {
			if attr, ok := l.ends[dst]; ok {
				ret[ln] = attr
			}
		}
	}
	return ret
}

// PurgeOutLinks removes outgoing edges from src that match the callback condition.
func (g *Graph) PurgeOutLinks(src string, cb PurgeOutLinkCallback, param interface{}) {
	g.mu.Lock()
	defer g.mu.Unlock()

	if n, ok := g.nodes[src]; ok {
		for ln, l := range n.outs {
			for dst, attr := range l.ends {
				if cb(src, ln, dst, attr, param) {
					delete(l.ends, dst)
					if len(l.ends) == 0 {
						delete(n.outs, ln)
					}
				}
			}
		}
	}
}

// GetEdgeCount returns the total number of edges in the graph.
func (g *Graph) GetEdgeCount() int {
	g.mu.RLock()
	defer g.mu.RUnlock()

	count := 0
	for _, n := range g.nodes {
		for _, l := range n.outs {
			count += len(l.ends)
		}
	}
	return count
}

// GraphSnapshot represents a serializable snapshot of the graph.
type GraphSnapshot struct {
	Nodes []string               `json:"nodes"`
	Edges []EdgeSnapshot         `json:"edges"`
}

// EdgeSnapshot represents a serializable edge.
type EdgeSnapshot struct {
	Src  string      `json:"src"`
	Link string      `json:"link"`
	Dst  string      `json:"dst"`
	Attr interface{} `json:"attr"`
}

// ToJSON serializes the graph to JSON.
func (g *Graph) ToJSON() ([]byte, error) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	snapshot := GraphSnapshot{
		Nodes: make([]string, 0, len(g.nodes)),
		Edges: make([]EdgeSnapshot, 0),
	}

	for node := range g.nodes {
		snapshot.Nodes = append(snapshot.Nodes, node)
	}

	for src, n := range g.nodes {
		for link, l := range n.outs {
			for dst, attr := range l.ends {
				snapshot.Edges = append(snapshot.Edges, EdgeSnapshot{
					Src:  src,
					Link: link,
					Dst:  dst,
					Attr: attr,
				})
			}
		}
	}

	return json.Marshal(snapshot)
}

// ForEachEdge iterates over all edges in the graph.
// The callback receives src, link type, dst, and attribute.
func (g *Graph) ForEachEdge(cb func(src, link, dst string, attr interface{})) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	for src, n := range g.nodes {
		for link, l := range n.outs {
			for dst, attr := range l.ends {
				cb(src, link, dst, attr)
			}
		}
	}
}

// ForEachNode iterates over all nodes in the graph.
func (g *Graph) ForEachNode(cb func(node string)) {
	g.mu.RLock()
	defer g.mu.RUnlock()

	for node := range g.nodes {
		cb(node)
	}
}
